syntax = "proto3";

package chord;

service NodeService {
    rpc Ping (PingRequest) returns (PingReply) {}
    rpc SearchFinger (FingerQuestion) returns (NodeInfoMessage) {}
    rpc NodeJoin (JoinRequest) returns (NodeInfoMessage) {}
    rpc Stabilize (NodeInfoMessage) returns (NodeInfoMessage) {}
    rpc Insert (InsertMessage) returns (NodeInfoMessage) {}
    rpc InsertMailbox (InsertMailboxMessage) returns (NodeInfoMessage) {}
    rpc Lookup (Query) returns (QueryResult) {}
    rpc LookupMailbox (QueryMailbox) returns (QueryResult) {}
}

message NodeInfoMessage {
    string ip = 1;
    int32 port = 2;
    int64 id = 3;
    bool cutoff = 4;
}

message FingerQuestion {
    int64 sender_id = 1;
    int64 finger_value = 2;
}

message JoinRequest {
    int64 node_id = 1;
}

message InsertMessage {
    int64 key = 1;
    string value = 2;
}

message InsertMailboxMessage {
    int64 key = 1;
    string owner = 2;
    int64 cutoff = 3;
}

message QueryMailbox {
    int64 key = 1;
}

message Query {
    int64 key = 1;
}

message QueryResult {
    NodeInfoMessage manager = 1;
    int64 key = 2;
    string value = 3;
}

message Empty { }

message PingRequest {
    uint64 ping_n = 1;
}

message PingReply {
    string ping_ip = 1;
    int32 ping_port = 2;
    int64 ping_id = 3;
    uint64 ping_n = 4;
}
